# 2.2 Answers

Filtering and Groupby Answers

_**Note:** This answers sheet is incomplete_

## Part 1

Run the following cells to import and clean the usgs data for the practice problems in part 1.

In [3]:
import random

import pandas as pd

In [4]:
# Read in the data
water_vars = pd.read_csv('../../1_python_progression/2-tabular_data/data/englewood_3_12_21_usgs_water.tsv', sep='\t', skiprows=30)
# There are a lot of variables here, so let's shorten our dataframe to a few variables
water_vars = water_vars[['datetime', '210920_00060', '210922_00010', '210924_00300', '210925_00400']]
# Get rid of the first row of hard-coded datatype info
water_vars = water_vars.drop(0)
# Rename the columns from their USGS codes to more human-readible names
name_codes = {'210920_00060': 'discharge','210922_00010': 'temperature', '210924_00300': 'dissolved oxygen', '210925_00400': 'pH'}
water_vars = water_vars.rename(columns=name_codes)
# Convert columns with numbers to a numeric type
water_vars['discharge'] = pd.to_numeric(water_vars['discharge'])
water_vars['temperature'] = pd.to_numeric(water_vars['temperature'])
water_vars['dissolved oxygen'] = pd.to_numeric(water_vars['dissolved oxygen'])
water_vars['pH'] = pd.to_numeric(water_vars['pH'])

water_vars['dam release'] = [random.choice([True, False]) for x in range(len(water_vars))]
water_vars['safety level'] = [random.choice(['low', 'medium', 'high']) for x in range(len(water_vars))]

### Filtering

1. Look at the output of `water_vars['safety level'] == 'high'`.  What is the output telling us?  Was the safety level high during the observation at index 3?

2. Create a dataframe showing which obersvations in `water_vars` recorded temperature less than 8.

3. Create a dataframe showing which obersvations in `water_vars` recorded either dissolved oxygen less than or equal to 8 or pH greater than 8.

4. Create a dataframe showing which obersvations in `water_vars` recorded discharge between 46 and 49.

### Named Indexes

The following bits of code add a new column to the dataframe made up of random letters and assigns them to a new column called `'sample id'`.  Run them before doing the next problems.

In [4]:
import string

In [5]:
letter_ids = [''.join(random.choice(string.ascii_lowercase) for i in range(5)) for x in range(146)]

In [6]:
water_vars['sample id'] = letter_ids

5. Use the new `'sample id'` column and set it as the index in the `water_vars` dataframe.

In [7]:
water_vars = water_vars.set_index('sample id')

In [8]:
water_vars

datetime  discharge  temperature  dissolved oxygen   pH  \
sample id                                                                    
dtsqs      2021-03-12 00:00       44.5          8.1               8.3  8.1   
jzsdq      2021-03-12 00:15       44.5          8.1               8.2  8.1   
bkncz      2021-03-12 00:30       44.5          8.1               8.2  8.1   
ieode      2021-03-12 00:45       44.5          8.1               8.1  8.1   
ebeds      2021-03-12 01:00       44.5          8.1               8.1  8.1   
...                     ...        ...          ...               ...  ...   
ribvb      2021-03-13 11:15       42.6          6.7               9.8  7.9   
epwyu      2021-03-13 11:30       42.6          6.7               9.9  7.9   
khqrz      2021-03-13 11:45       42.6          6.7              10.2  7.9   
dmqyc      2021-03-13 12:00       46.5          6.7              10.3  7.9   
rqmky      2021-03-13 12:15        NaN          6.6              10.3  7.9   

           dam release safety level  
sample id                            
dtsqs             True       medium  
jzsdq            False       medium  
bkncz            False         high  
ieode             True       medium  
ebeds            False          low  
...                ...          ...  
ribvb            False         high  
epwyu             True         high  
khqrz            False       medium  
dmqyc            False         high  
rqmky            False         high  

[146 rows x 7 columns]

6. Pick a sample id and get the data for just that row.

In [9]:
# print out your dataframe to pick an index that exists
# water_vars.loc['ceule']

7. We have mostly been looking at boolean comparisons with numbers, but you can also interrogate strings with special methods and get a list of booleans as a result.  Consider the following bit of code:

In [10]:
water_vars.index.str.contains('k')

array([False, False,  True, False, False, False,  True,  True, False,
        True, False, False,  True, False, False, False, False, False,
       False, False,  True,  True, False,  True, False, False,  True,
       False,  True, False, False,  True,  True, False, False, False,
       False, False, False, False, False, False, False,  True, False,
       False, False, False, False, False, False, False, False, False,
        True, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
        True,  True, False, False, False, False, False, False, False,
       False, False, False, False, False, False,  True, False, False,
       False, False, False,  True, False, False, False, False, False,
       False, False,  True, False,  True, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False,  True,

This code is telling us True/False for which indexes contain the letter k in them.

Using what you know about pandas filtering, use `water_vars.loc[]` and the code above (`water_vars.index.str.contains('k')`) to filter the data frame to only return the rows which have an index containing the letter k.

In [11]:
water_vars.loc[water_vars.index.str.contains('k')]

datetime  discharge  temperature  dissolved oxygen   pH  \
sample id                                                                    
bkncz      2021-03-12 00:30       44.5          8.1               8.2  8.1   
bwnkw      2021-03-12 01:30       44.5          8.0               8.0  8.1   
cozgk      2021-03-12 01:45       44.5          8.0               8.0  8.1   
kfgfq      2021-03-12 02:15       44.5          8.0               8.0  8.1   
klccb      2021-03-12 03:00       44.5          7.8               8.0  8.1   
gtrik      2021-03-12 05:00       46.5          7.4               8.0  8.0   
snkxq      2021-03-12 05:15       46.5          7.4               8.0  8.0   
lykrn      2021-03-12 05:45       48.5          7.3               8.1  8.0   
kxdas      2021-03-12 06:30       48.5          7.1               8.2  7.9   
qkdcf      2021-03-12 07:00       48.5          7.0               8.2  7.9   
firuk      2021-03-12 07:45       48.5          6.8               8.3  7.9   
knlyk      2021-03-12 08:00       48.5          6.8               8.4  7.9   
ktbjf      2021-03-12 10:45       44.5          6.8               9.8  7.9   
uzxkm      2021-03-12 13:30       40.8          7.6              11.9  8.1   
xyypk      2021-03-12 20:15       46.5          8.1              10.2  7.9   
bmrik      2021-03-12 20:30       46.5          8.1              10.0  7.9   
lkvwz      2021-03-13 00:00       44.5          7.7               8.6  8.0   
mkmid      2021-03-13 01:30       44.5          7.7               8.4  8.0   
bgkpx      2021-03-13 03:30       44.5          7.6               8.2  8.0   
zikfz      2021-03-13 04:00       44.5          7.5               8.2  8.0   
omqok      2021-03-13 07:45       46.5          6.9               8.4  7.8   
kfxir      2021-03-13 09:15       46.5          6.8               9.0  7.9   
pgfyk      2021-03-13 10:45       44.5          6.7               9.6  7.9   
knsjs      2021-03-13 11:00       42.6          6.7               9.7  7.9   
khqrz      2021-03-13 11:45       42.6          6.7              10.2  7.9   
rqmky      2021-03-13 12:15        NaN          6.6              10.3  7.9   

           dam release safety level  
sample id                            
bkncz            False         high  
bwnkw            False          low  
cozgk            False          low  
kfgfq            False       medium  
klccb             True         high  
gtrik            False          low  
snkxq            False          low  
lykrn             True       medium  
kxdas             True          low  
qkdcf             True       medium  
firuk             True       medium  
knlyk             True          low  
ktbjf            False         high  
uzxkm            False       medium  
xyypk             True         high  
bmrik             True         high  
lkvwz            False         high  
mkmid            False       medium  
bgkpx             True          low  
zikfz            False          low  
omqok            False       medium  
kfxir            False       medium  
pgfyk            False         high  
knsjs             True         high  
khqrz            False       medium  
rqmky            False         high

### Groupby

8. Fill in the three groupby questions for the following scenario:

**Given the environmental sustainbility index (ESI) for every country, what is the minimum ESI on each continent?**

* _Which variable to group together?_
* _How do we want to group?_
* _What variable do we want to look at?_ 

8. Fill in the three groupby questions for the following scenario:

**Given the graduation rates of many high school, what is the average graduation rate in public vs. private schools?**

* _Which variable to group together?_ 
* _How do we want to group?_
* _What variable do we want to look at?_ 

9. Find the mean values of all of the columns at each safety level

In [12]:
water_vars.groupby("safety level").mean()

discharge  temperature  dissolved oxygen        pH  dam release
safety level                                                                 
high          44.845283     7.422222          9.390741  7.972222     0.481481
low           44.731429     7.577143          9.388571  8.008571     0.485714
medium        44.810526     7.560000          9.426316  7.991228     0.403509

10. Find the mean values of all of the columns at each discharge (less meaningful, but just for practice)

In [13]:
water_vars.groupby("discharge").mean()

temperature  dissolved oxygen        pH  dam release
discharge                                                      
39.0          8.385714         12.957143  8.200000     0.571429
40.8          8.027273         12.309091  8.136364     0.272727
42.6          7.261538         10.807692  7.992308     0.461538
44.5          7.665385          8.778846  8.005769     0.480769
46.5          7.314583          8.820000  7.922000     0.360000
48.5          7.008333          8.216667  7.933333     0.833333

11. Find the number of non-nan observations for each column at each safety level

In [21]:
water_vars.groupby("safety level").count()

datetime  discharge  temperature  dissolved oxygen  pH  \
safety level                                                           
high                46         46           46                46  46   
low                 53         53           51                53  53   
medium              47         46           47                47  47   

              dam release  
safety level               
high                   46  
low                    53  
medium                 47

12. Display the mean values of just the temperature column at each safety level

In [22]:
water_vars.groupby("safety level")['temperature'].mean()

safety level
high      7.426087
low       7.590196
medium    7.512766
Name: temperature, dtype: float64

## Part 2

For the next set of practice problems we are going to use the WAS 2020 SARP data.  Run the following lines of code to import the data.

In [14]:
was_2020_filepath = "../data/SARP 2020 final.xlsx"
was_2020 = pd.read_excel(was_2020_filepath, "INPUT", skipfooter=7)

### Question 1

**A)** Using `was_2020` find data samples where `'CO2 (ppmv)'` was greater than 420.

In [25]:
was_2020[was_2020['CO2 (ppmv)'] > 420]

Can #  Snake can #                                          Location  \
1    7338         3716                  Buffalo Park, Flagstaff, Arizona   
3    5012         3702                  Buffalo Park, Flagstaff, Arizona   
5    5061         3703                          Downtown Sedona, Arizona   
13   6123         3721  Mars Hill/Lowell Observatory, Flagstaff, Arizona   
15   7335         3711                          Downtown Sedona, Arizona   
..    ...          ...                                               ...   
898  9172         2211    Elizabeth Park, central Bellingham, Washington   
902  9168         2208         My backyard, rural Bellingham, Washington   
905  5109         2601    Carl Traeger middle school, Oshkosh, Wisconsin   
906  7027         2616                   South Park, Oshkosh, Wisconsin    
908  5131         2602                   South Park, Oshkosh, Wisconsin    

          State  Box       Date     Time  Altitude (m) Temp. Celsius  \
1       Arizona   37 2020-04-24  11:44:00         2100            18   
3       Arizona   37 2020-05-01  11:16:00         2100            19   
5       Arizona   37 2020-05-01  12:40:00         1350            26   
13      Arizona   37 2020-06-03  12:50:00         2200            27   
15      Arizona   37 2020-06-21  17:20:00         1350            36   
..          ...  ...        ...       ...          ...           ...   
898  Washington   22 2020-06-05  10:59:00           22          16.1   
902  Washington   22 2020-07-15  10:00:00          153          17.2   
905   Wisconsin   26 2020-04-15  10:49:00       243.84          -2.2   
906   Wisconsin   26 2020-04-15  11:01:00       231.64          -1.1   
908   Wisconsin   26 2020-04-22  12:26:00       231.64           7.2   

    Wind Spd (mph)  ... 3-Ethyltoluene (MS) 4-Ethyltoluene (MS)  \
1                5  ...                  12                  16   
3               16  ...                -888                -888   
5                8  ...                -888                   3   
13               8  ...                   4                -888   
15              12  ...                  34                  15   
..             ...  ...                 ...                 ...   
898              5  ...                   5                -888   
902         0-2mph  ...                   4                   3   
905              7  ...                  55                  25   
906              3  ...                -888                -888   
908             17  ...                -888                -888   

    2-Ethyltoluene (MS)  1,3,5-Trimethylbenzene (MS)  \
1                    15                           14   
3                  -888                         -888   
5                  -888                         -888   
13                 -888                         -888   
15                   13                           16   
..                  ...                          ...   
898                -888                         -888   
902                   3                         -888   
905                  17                           19   
906                -888                         -888   
908                -888                         -888   

     1,2,4-Trimethylbenzene (MS)  alpha-Pinene (B/MS)  beta-Pinene (B/MS)  \
1                             22                   31                   9   
3                           -888                   22                -888   
5                              5                   26                -888   
13                             6                   67                  16   
15                            51                   16                -888   
..                           ...                  ...                 ...   
898                            7                    5                -888   
902                            8                  121                  26   
905                           64                   47  

**B)** How many data samples had CO2 over 420?

### Question 2

**A)** What is the mean value of `'Isoprene (E/B)'` accross all the samples?

In [26]:
was_2020['Isoprene (E/B)'].mean()

346.80193756727664

**B)** How many dta samples had values of 'Isoprene (E/B)' below 10?

In [27]:
was_2020[was_2020['Isoprene (E/B)'] < 10]['Isoprene (E/B)']

0        4
26       5
27    -888
35       5
66       6
      ... 
911      5
912   -888
915      5
917      4
920      4
Name: Isoprene (E/B), Length: 157, dtype: int64

**C)** Print out the values of isoprene from the previous part of the question which are less than 10.  Do you notice anything unusual about the values?

**D)** Looking at the values in the previous question you might see that several values are marked with -888.  This is likely a nodata value (but its always good to check the data documentation to be sure).

Knowing that there are now nodata values, go back and ammend your code to find the rows where isoprene is less than 10 but also not equal to -888.  Now how many rows are there?

**E)** What is the true mean value of isoprene?

_This question was brought to you today as a gentle reminder to always be on the lookout for nodata 🙂_

### Question 3

There are a lot of data columns in this dataset, many more than we can reasonably view in the preview.  To help us explore our data better we might want to view a smaller subset of the data at once.

**A)** Create a dataframe that includes only the following columns from the `was_2020` dataset: `'Can #'`, `'State'`, `'Date'`, `'CO (ppbv)'`, and `'OCS (MS)'`.

**B)** Let's say you would like to view all of the columns of CFCs together.  There are several columns that all start with `CFC-`: 'CFC-12 (C/D)', 'CFC-11 (D/MS)', 'CFC-113 (D)', 'CFC-114 (MS)'.

Create a dataframe that includes just the columns whose names start with "CFC-".

_**Google help:** try "pandas filter by column name string".  [This stackoverflow](https://stackoverflow.com/questions/21285380/find-column-whose-name-contains-a-specific-string) also has several options.  The first few use a method called "list comprehension" that we haven't talked about in lecture yet.  Further down are a few other options to try that don't use that technique._

**C)** Use the subset of data you created in part be and return just the columns of CFCs and only the rows of samples that were taken in Arizona.

### Question 4 - `groupby` questions

**A)** What are the mean values of `'CO (ppbv)'` from each box?



In [28]:
was_2020.groupby('Box')['CO (ppbv)'].mean()

Box
1     128.652174
3     141.217391
4     125.565217
6     114.333333
7     138.695652
8     204.454545
9     182.045455
10    122.217391
12    240.041667
14    121.818182
15    136.850000
16    127.083333
17    147.826087
19    137.863636
22    108.695652
23    151.958333
26    119.625000
27    157.739130
28    150.391304
29    152.304348
30    115.909091
31    136.125000
32    120.565217
34    118.380952
35    154.318182
36    159.541667
37    111.727273
38    120.285714
40    121.652174
41    197.045455
46    155.136364
49    102.125000
50    132.363636
52    102.625000
54    101.400000
55    107.521739
56    155.590909
57    143.000000
60    112.333333
61    110.260870
62    133.391304
Name: CO (ppbv), dtype: float64

**B)** What is the minimum value of `'CH4 (ppmv height)'` for each `Weather` condition?

In [29]:
was_2020.groupby('Weather')['CH4 (ppmv height)'].min()

Weather
Clear                            1.892
Clear and sunny                  1.882
Clear skies and sunny            1.936
Clear skies, sunny, warm wind    1.933
Clear/Sunny                      1.976
                                 ...  
sunny sky but hazy               1.980
sunny, but a bit hazy            1.950
sunny, hazy                      1.993
sunny, slightly hazy             1.976
windy, about to rain             1.932
Name: CH4 (ppmv height), Length: 87, dtype: float64

**C)** How many samples were taken in each state?

In [30]:
was_2020.groupby("State")["Can #"].count()

State
Arizona           22
California       204
Colorado          64
Connecticut       23
Florida           22
Georgia           23
Illinois          22
Kentucky          24
Louisiana         21
Massachusetts     46
Minnesota         24
Missouri          22
New Jersey        23
New York          94
Ohio              22
Oregon            24
Texas            113
Utah              23
Virginia          66
Washington        23
Wisconsin         24
Name: Can #, dtype: int64

_For the final two questions we will switch back to the water dataset_

### Question 5

`.groupby` also allows you to group on multiple fields.  So instead of saying "what is the mean value of CO in each state?", we can be more specific, and ask, "What is the mean value of CO in each state at every wind speed?"

We do this by using a `list` of values in the `.groupby()` argument instead of just a single column string.

`.groupby(["Box", "Wind Spd (mph)"])`

**A)** Try using this syntax to find the mean value of CO in each state at every wind speed

In [31]:
was_2020.groupby(["State", "Wind Spd (mph)"])['CO (ppbv)'].mean()['Illinois']

Wind Spd (mph)
3        128.0
4        124.5
5        229.5
6        122.5
7        151.0
8        153.0
9         79.0
10        97.0
11       115.0
15-25     94.0
15-26    108.0
15-27    116.0
17-26    109.0
18-26    111.0
Name: CO (ppbv), dtype: float64

**B)** What was the mean value of CO in Georgia when the wind speed was 8 mph?

### Question 6

The current `.groupby` function allows to use a single aggregation function at once -- we can see either the means for every columns, or the max of every column, and so on.

To look use different aggregation functions for different columns we drop the aggregation function and instead use `.agg()`.

Here is an example using the water dataset:

In [32]:
water_vars.groupby("dam release").agg(
    max_discharge=pd.NamedAgg(column='discharge', aggfunc='max'), 
    mean_doxy=pd.NamedAgg(column="dissolved oxygen", aggfunc='mean')
)

max_discharge  mean_doxy
dam release                          
False                 48.5   9.466667
True                  48.5   9.343243

Use the `.agg()` function to find the maximum values of `'CH4 (ppmv height)'`, the mean values of 'Toluene (E/B/MS)', and the total number of samples for each state.

In [34]:
# copied improperly
# was_2020.groupby("State").agg(
#     max_discharge=pd.NamedAgg(column='discharge', aggfunc='max'), 
#     mean_doxy=pd.NamedAgg(column="dissolved oxygen", aggfunc='mean')
# )